File: src/glimpse.mjs
import { EventEmitter } from 'node:events';
import { spawn } from 'node:child_process';
import { createInterface } from 'node:readline';
import { existsSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const BINARY = join(__dirname, 'glimpse');

class GlimpseWindow extends EventEmitter {
  #proc;
  #closed = false;
  #pendingHTML = null;

  constructor(proc, initialHTML) {
    super();
    this.#proc = proc;
    this.#pendingHTML = initialHTML;

    proc.stdin.on('error', () => {}); // Swallow EPIPE if Swift exits first

    const rl = createInterface({ input: proc.stdout, crlfDelay: Infinity });

    rl.on('line', (line) => {
      let msg;
      try {
        msg = JSON.parse(line);
      } catch {
        this.emit('error', new Error(`Malformed protocol line: ${line}`));
        return;
      }

      switch (msg.type) {
        case 'ready':
          if (this.#pendingHTML) {
            // First ready = blank page loaded. Send the queued HTML.
            this.setHTML(this.#pendingHTML);
            this.#pendingHTML = null;
          } else {
            // Subsequent ready = user HTML loaded. Notify caller.
            this.emit('ready');
          }
          break;
        case 'message':
          this.emit('message', msg.data);
          break;
        case 'closed':
          if (!this.#closed) {
            this.#closed = true;
            this.emit('closed');
          }
          break;
        default:
          break;
      }
    });

    proc.on('error', (err) => this.emit('error', err));

    proc.on('exit', () => {
      if (!this.#closed) {
        this.#closed = true;
        this.emit('closed');
      }
    });
  }

  #write(obj) {
    if (this.#closed) return;
    this.#proc.stdin.write(JSON.stringify(obj) + '\n');
  }

  send(js) {
    this.#write({ type: 'eval', js });
  }

  setHTML(html) {
    this.#write({ type: 'html', html: Buffer.from(html).toString('base64') });
  }

  close() {
    this.#write({ type: 'close' });
  }

  loadFile(path) {
    this.#write({ type: 'file', path });
  }

  followCursor(enabled) {
    this.#write({ type: 'follow-cursor', enabled });
  }
}

export function open(html, options = {}) {
  if (!existsSync(BINARY)) {
    throw new Error(
      "Glimpse binary not found. Run 'npm run build' or 'swiftc src/glimpse.swift -o src/glimpse'"
    );
  }

  const args = [];
  if (options.width != null)  args.push('--width',  String(options.width));
  if (options.height != null) args.push('--height', String(options.height));
  if (options.title != null)  args.push('--title',  options.title);

  if (options.frameless)    args.push('--frameless');
  if (options.floating)     args.push('--floating');
  if (options.transparent)  args.push('--transparent');
  if (options.clickThrough) args.push('--click-through');
  if (options.followCursor) args.push('--follow-cursor');
  if (options.autoClose)   args.push('--auto-close');

  if (options.x != null) args.push('--x', String(options.x));
  if (options.y != null) args.push('--y', String(options.y));

  if (options.cursorOffset?.x != null) args.push('--cursor-offset-x', String(options.cursorOffset.x));
  if (options.cursorOffset?.y != null) args.push('--cursor-offset-y', String(options.cursorOffset.y));

  const proc = spawn(BINARY, args, { stdio: ['pipe', 'pipe', 'inherit'] });
  return new GlimpseWindow(proc, html);
}

export function prompt(html, options = {}) {
  return new Promise((resolve, reject) => {
    const win = open(html, { ...options, autoClose: true });
    let resolved = false;

    const timer = options.timeout
      ? setTimeout(() => {
          if (!resolved) { resolved = true; win.close(); reject(new Error('Prompt timed out')); }
        }, options.timeout)
      : null;

    win.once('message', (data) => {
      if (!resolved) {
        resolved = true;
        if (timer) clearTimeout(timer);
        resolve(data);
      }
    });

    win.once('closed', () => {
      if (timer) clearTimeout(timer);
      if (!resolved) {
        resolved = true;
        resolve(null); // User closed window without sending a message
      }
    });

    win.once('error', (err) => {
      if (timer) clearTimeout(timer);
      if (!resolved) {
        resolved = true;
        reject(err);
      }
    });
  });
}

File: src/glimpse.swift
import Cocoa
import WebKit
import Foundation

// MARK: - Stdout Helper

func writeToStdout(_ dict: [String: Any]) {
    guard let data = try? JSONSerialization.data(withJSONObject: dict),
          let line = String(data: data, encoding: .utf8) else { return }
    let output = line + "\n"
    FileHandle.standardOutput.write(output.data(using: .utf8)!)
    fflush(stdout)
}

func log(_ message: String) {
    fputs("[glimpse] \(message)\n", stderr)
}

// MARK: - System Info

func getSystemInfo() -> [String: Any] {
    let mouse = NSEvent.mouseLocation

    // Main screen
    var screenInfo: [String: Any] = [:]
    if let screen = NSScreen.main {
        let f = screen.frame
        let v = screen.visibleFrame
        screenInfo = [
            "width": Int(f.width),
            "height": Int(f.height),
            "scaleFactor": Int(screen.backingScaleFactor),
            "visibleX": Int(v.origin.x),
            "visibleY": Int(v.origin.y),
            "visibleWidth": Int(v.width),
            "visibleHeight": Int(v.height),
        ]
    }

    // All screens
    let screens: [[String: Any]] = NSScreen.screens.map { screen in
        let f = screen.frame
        let v = screen.visibleFrame
        return [
            "x": Int(f.origin.x),
            "y": Int(f.origin.y),
            "width": Int(f.width),
            "height": Int(f.height),
            "scaleFactor": Int(screen.backingScaleFactor),
            "visibleX": Int(v.origin.x),
            "visibleY": Int(v.origin.y),
            "visibleWidth": Int(v.width),
            "visibleHeight": Int(v.height),
        ]
    }

    // Appearance
    let isDark = NSApp.effectiveAppearance.bestMatch(from: [.darkAqua, .aqua]) == .darkAqua
    let accent = NSColor.controlAccentColor.usingColorSpace(.sRGB)
    let accentHex: String
    if let c = accent {
        accentHex = String(format: "#%02X%02X%02X", Int(c.redComponent * 255), Int(c.greenComponent * 255), Int(c.blueComponent * 255))
    } else {
        accentHex = "#007AFF"
    }
    let reduceMotion = NSWorkspace.shared.accessibilityDisplayShouldReduceMotion
    let increaseContrast = NSWorkspace.shared.accessibilityDisplayShouldIncreaseContrast

    return [
        "screen": screenInfo,
        "screens": screens,
        "appearance": [
            "darkMode": isDark,
            "accentColor": accentHex,
            "reduceMotion": reduceMotion,
            "increaseContrast": increaseContrast,
        ],
        "cursor": [
            "x": Int(mouse.x),
            "y": Int(mouse.y),
        ],
    ]
}

// MARK: - CLI Config

struct Config {
    var width: Int = 800
    var height: Int = 600
    var title: String = "Glimpse"
    var frameless: Bool = false
    var floating: Bool = false
    var transparent: Bool = false
    var x: Int? = nil
    var y: Int? = nil
    var followCursor: Bool = false
    var cursorOffsetX: Int = 20
    var cursorOffsetY: Int = -20
    var clickThrough: Bool = false
    var autoClose: Bool = false
}

func parseArgs() -> Config {
    var config = Config()
    let args = CommandLine.arguments
    var i = 1
    while i < args.count {
        switch args[i] {
        case "--width":
            i += 1
            if i < args.count, let v = Int(args[i]) { config.width = v }
        case "--height":
            i += 1
            if i < args.count, let v = Int(args[i]) { config.height = v }
        case "--title":
            i += 1
            if i < args.count { config.title = args[i] }
        case "--frameless":
            config.frameless = true
        case "--floating":
            config.floating = true
        case "--transparent":
            config.transparent = true
        case "--x":
            i += 1
            if i < args.count, let v = Int(args[i]) { config.x = v }
        case "--y":
            i += 1
            if i < args.count, let v = Int(args[i]) { config.y = v }
        case "--follow-cursor":
            config.followCursor = true
        case "--cursor-offset-x":
            i += 1
            if i < args.count, let v = Int(args[i]) { config.cursorOffsetX = v }
        case "--cursor-offset-y":
            i += 1
            if i < args.count, let v = Int(args[i]) { config.cursorOffsetY = v }
        case "--click-through":
            config.clickThrough = true
        case "--auto-close":
            config.autoClose = true
        default:
            break
        }
        i += 1
    }
    return config
}

// MARK: - Window Subclass (keyboard support for frameless windows)

class GlimpsePanel: NSWindow {
    override var canBecomeKey: Bool { true }
    override var canBecomeMain: Bool { true }
}

// MARK: - AppDelegate

@MainActor
class AppDelegate: NSObject, NSApplicationDelegate, WKNavigationDelegate, WKScriptMessageHandler, NSWindowDelegate {

    var window: NSWindow!
    var webView: WKWebView!
    let config: Config

    // Mouse monitor references for follow-cursor mode
    var globalMouseMonitor: Any?
    var localMouseMonitor: Any?

    nonisolated init(config: Config) {
        self.config = config
    }

    func applicationDidFinishLaunching(_ notification: Notification) {
        setupWindow()
        setupWebView()
        if config.followCursor {
            startFollowingCursor()
        }
        startStdinReader()
    }

    // MARK: - Setup

    private func setupWindow() {
        let rect = NSRect(x: 0, y: 0, width: config.width, height: config.height)
        let styleMask: NSWindow.StyleMask = config.frameless
            ? [.borderless]
            : [.titled, .closable, .miniaturizable, .resizable]
        window = GlimpsePanel(
            contentRect: rect,
            styleMask: styleMask,
            backing: .buffered,
            defer: false
        )
        window.title = config.title
        if config.frameless {
            window.isMovableByWindowBackground = true
        }
        if config.floating || config.followCursor {
            window.level = .floating
        }
        if config.clickThrough {
            window.ignoresMouseEvents = true
        }
        if config.transparent {
            window.isOpaque = false
            window.backgroundColor = .clear
        }
        if config.followCursor {
            let mouse = NSEvent.mouseLocation
            let x = mouse.x + CGFloat(config.cursorOffsetX)
            let y = mouse.y + CGFloat(config.cursorOffsetY)
            window.setFrameOrigin(NSPoint(x: x, y: y))
        } else if let x = config.x, let y = config.y {
            window.setFrameOrigin(NSPoint(x: x, y: y))
        } else {
            window.center()
        }
        window.delegate = self
        if config.clickThrough {
            window.orderFrontRegardless()
        } else {
            window.makeKeyAndOrderFront(nil)
            NSApp.activate(ignoringOtherApps: true)
        }
    }

    private func setupWebView() {
        let ucc = WKUserContentController()

        let bridgeJS = """
        window.glimpse = {
            send: function(data) {
                window.webkit.messageHandlers.glimpse.postMessage(JSON.stringify(data));
            },
            close: function() {
                window.webkit.messageHandlers.glimpse.postMessage(JSON.stringify({__glimpse_close: true}));
            }
        };
        """
        let script = WKUserScript(source: bridgeJS, injectionTime: .atDocumentStart, forMainFrameOnly: true)
        ucc.addUserScript(script)
        ucc.add(self, name: "glimpse")

        let wkConfig = WKWebViewConfiguration()
        wkConfig.userContentController = ucc

        webView = WKWebView(frame: window.contentView!.bounds, configuration: wkConfig)
        webView.autoresizingMask = [.width, .height]
        webView.navigationDelegate = self
        if config.transparent {
            webView.underPageBackgroundColor = .clear
            webView.setValue(false, forKey: "drawsBackground")
        }
        window.contentView?.addSubview(webView)

        // Load blank page so didFinish fires and we emit "ready"
        webView.loadHTMLString("<html><body></body></html>", baseURL: nil)
    }

    // MARK: - Follow Cursor

    func startFollowingCursor() {
        guard globalMouseMonitor == nil else { return }
        window.level = .floating
        let moveHandler: (NSEvent) -> Void = { [weak self] _ in
            guard let self else { return }
            let mouse = NSEvent.mouseLocation
            let x = mouse.x + CGFloat(self.config.cursorOffsetX)
            let y = mouse.y + CGFloat(self.config.cursorOffsetY)
            self.window.setFrameOrigin(NSPoint(x: x, y: y))
        }
        globalMouseMonitor = NSEvent.addGlobalMonitorForEvents(
            matching: [.mouseMoved, .leftMouseDragged, .rightMouseDragged],
            handler: moveHandler
        )
        localMouseMonitor = NSEvent.addLocalMonitorForEvents(matching: [.mouseMoved, .leftMouseDragged, .rightMouseDragged]) { [weak self] event in
            guard let self else { return event }
            let mouse = NSEvent.mouseLocation
            let x = mouse.x + CGFloat(self.config.cursorOffsetX)
            let y = mouse.y + CGFloat(self.config.cursorOffsetY)
            self.window.setFrameOrigin(NSPoint(x: x, y: y))
            return event
        }
    }

    func stopFollowingCursor() {
        if let monitor = globalMouseMonitor {
            NSEvent.removeMonitor(monitor)
            globalMouseMonitor = nil
        }
        if let monitor = localMouseMonitor {
            NSEvent.removeMonitor(monitor)
            localMouseMonitor = nil
        }
    }

    // MARK: - Stdin Reader

    private func startStdinReader() {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            while let line = readLine() {
                let trimmed = line.trimmingCharacters(in: .whitespaces)
                guard !trimmed.isEmpty else { continue }
                guard let data = trimmed.data(using: .utf8),
                      let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let type = json["type"] as? String
                else {
                    log("Skipping invalid JSON: \(trimmed)")
                    continue
                }
                DispatchQueue.main.async {
                    MainActor.assumeIsolated {
                        self?.handleCommand(type: type, json: json)
                    }
                }
            }
            // stdin EOF — close window
            DispatchQueue.main.async {
                MainActor.assumeIsolated {
                    self?.closeAndExit()
                }
            }
        }
    }

    // MARK: - Command Dispatch

    func handleCommand(type: String, json: [String: Any]) {
        switch type {
        case "html":
            guard let base64 = json["html"] as? String,
                  let htmlData = Data(base64Encoded: base64),
                  let html = String(data: htmlData, encoding: .utf8)
            else {
                log("html command: missing or invalid base64 payload")
                return
            }
            webView.loadHTMLString(html, baseURL: nil)
        case "eval":
            guard let js = json["js"] as? String else {
                log("eval command: missing js field")
                return
            }
            webView.evaluateJavaScript(js, completionHandler: nil)
        case "follow-cursor":
            let enabled = json["enabled"] as? Bool ?? true
            if enabled {
                startFollowingCursor()
            } else {
                stopFollowingCursor()
            }
        case "file":
            guard let path = json["path"] as? String else {
                log("file command: missing path field")
                return
            }
            let fileURL = URL(fileURLWithPath: path)
            guard FileManager.default.fileExists(atPath: path) else {
                log("file command: file not found: \(path)")
                return
            }
            webView.loadFileURL(fileURL, allowingReadAccessTo: fileURL.deletingLastPathComponent())
        case "get-info":
            var info = getSystemInfo()
            info["type"] = "info"
            writeToStdout(info)
        case "close":
            closeAndExit()
        default:
            log("Unknown command type: \(type)")
        }
    }

    func closeAndExit() {
        writeToStdout(["type": "closed"])
        exit(0)
    }

    // MARK: - WKNavigationDelegate

    nonisolated func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        MainActor.assumeIsolated {
            window.makeFirstResponder(webView)
            var info = getSystemInfo()
            info["type"] = "ready"
            writeToStdout(info)
        }
    }

    // MARK: - WKScriptMessageHandler

    nonisolated func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        MainActor.assumeIsolated {
            guard let body = message.body as? String,
                  let data = body.data(using: .utf8),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else {
                log("Received invalid message from webview")
                return
            }

            if json["__glimpse_close"] as? Bool == true {
                closeAndExit()
                return
            }

            writeToStdout(["type": "message", "data": json])
            if config.autoClose {
                closeAndExit()
            }
        }
    }

    // MARK: - NSWindowDelegate

    func windowWillClose(_ notification: Notification) {
        writeToStdout(["type": "closed"])
        exit(0)
    }
}

// MARK: - Entry Point

let config = parseArgs()
let app = NSApplication.shared
let delegate = AppDelegate(config: config)
app.delegate = delegate
app.setActivationPolicy(config.clickThrough ? .accessory : .regular)
app.run()
